Improve Kotlin tour (#4321)
JetBrains/kotlin-web-site: docs/topics/tour/kotlin-tour-functions.md

Reviewer: I think the section about the lambdas reads much better now. I would suggest to add an example of passing the lambda defined separately to a function.

```kotlin
val numbers = listOf(1, -2, 3, -4, 5, -6)
val isPositive: (Int) -> Boolean = { x -> x > 0 }
val positives = numbers.filter(isPositive)
```

Author: Unfortunately we don't cover function types until after this section so I can't add this here 😞

Author: Fixed after slight adjustment to the example!

```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3" id="kotlin-tour-simple-function-before"}

You can remove the curly braces `{}` and declare the function body using the assignment operator `=`. When you use the 

> assignment operator `=`, Kotlin uses type inference, so you can also omit the return type. The `sum()` function then becomes one line:

Reviewer: ```suggestion
assignment operator `=`, Kotlin uses type inference, so you can also omit the return type.
(although it is still highly recommended to write it explicitly for public code.)
The `sum()` function then becomes one line:
```

Reviewer: I was going to include a link to the [author guidelines](https://kotlinlang.org/docs/api-guidelines-introduction.html) but I then thought that may be too much for an introduction to the language.

Author: As discussed, yeah it's a bit heavy a this stage but I will add the tip with slightly different wording!

```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3" id="kotlin-tour-simple-function-after"}

However, if you want your code to be quickly understood by other developers, it's a good idea to explicitly define the 
return type even when using the assignment operator `=`.

> > If you use `{}` curly braces to declare your function body, you must declare the return type unless it is `Unit`.

Reviewer 2: ```suggestion
> If you use `{}` curly braces to declare your function body, you must declare the return type unless it's a `Unit`.
```

Author: What about "it is `Unit` type"?

Reviewer 2: OK, but shouldn't it be "it is the `Unit` type" in this case?

* `text` has type `String`.
* the function returns the result of the [`.uppercase()`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/uppercase.html)
function called on `text`.

> * the entire lambda expression is assigned to the `upperCaseString` variable with the assignment operator `=`.

Reviewer 2: I'm surprised to see you start list items almost everywhere with a lowercase letter. I don't think it's an option in the guide and it's not consistent with the rest of the documentation

* `{ x -> x > 0 }` takes each element of the list and returns only those that are positive.
* `{ x -> x < 0 }` takes each element of the list and returns only those that are negative.

> This example demonstrates two ways of passing a lambda expression to a function: 

Reviewer: I like this, it shows very clearly how function arguments are not special at all (except for the trailing lambda syntax).
